students/kyle_odonnell/Lesson04/mailroom.py

```diff
@@ -0,0 +1,258 @@
+    :param db: (list) of dictionaries with donor information
+    :return: nothing
+    """
+    db.sort(key=get_total_given, reverse=True)  # Sort table by total donated amount
+    heading = "| {dn:<20s}\t| {tg:<10s}\t| {ng:<10s} | {ag:<10s}   |".format
+    row = "{dn:<20s} \t {ds:<1s} {tg:>9.2f} \t {ng:>10d} \t {ds2:<1} {ag:>9.2f} ".format
+    print(heading(dn="Donor Name", tg="Total Given", ng="Num Gifts", ag="Average Gift"))
+    print("-----------------------------------------------------------------------")
+    for entry in db:
+        print(row(dn=entry["Donor Name"], ds="$", tg=entry["Total Given"],
+                  ng=entry["Num Gifts"], ds2="$", ag=entry["Average Gift"]))
+    return list(db)
+
+
+def welcome_message():  # Display welcome message
+    """ Print welcome message when program is launched
+    :return: nothing
+    """
+    print("\nWelcome to the Mailroom Program!")
+
+
+def close_app():  # Display message when user leaves
+    """ Print goodbye message when user closes program
+    :return: nothing
+    """
+    print("Closing Mailroom App...Goodbye!")
+
+
+def option_one_func(db):  # Update database
+    """ Add new name or new donation to existing name to donor database
+    :param db: (list) of dictionaries with donor information
+    :return: nothing
+    """
+    while True:
+        name_string = enter_name()
+        if name_string.lower() == "cancel":  # Cancel task
+            break
+        elif name_string.lower().strip() == "list":  # Print donor names
+            display_names(db)
+        else:
+            try:
+                donation_amount = enter_donation(name_string)  # Prompt donation amount
+                db = add_donation_amount(name_string, donation_amount, db)
+                thank_you_letter(name_string, donation_amount)
+                break
+            except ValueError:
+                print("Entry failed: Donations must be entered as a number!")
+                break
+    return list(db)
+
+
+def generate_letter_choice(db):
+    """ Write letters to text files for every donor in database
+    :param db: (list) of dictionaries with donation information:
+    :return: nothing
+    """
+    file_option = int(input("""
+    Okay, let's create some letters! You can:
+    1. Create a letter for a specific donor
+    2. Create letters for all donors \n
+    Please select an option (1 or 2): """))
+    db = switch_func_letter_dict.get(file_option)(db)
+    return db
+
+
+switch_func_dict = {
+    1: option_one_func,
+    2: create_report,
+    3: generate_letter_choice
+}
+
+
+switch_func_letter_dict = {
+    1: generate_one_letter,
+    2: generate_all_letters
+}
+
+# Main Body of Script  ---------------------------------------------------- #
+if __name__ == '__main__':
+    welcome_message()
+    while True:
+        menu_options()
+        try:
+            choice = int(prompt_menu_option())
+            if choice == 4:
+                break
```

Comment:
you should map this to a function as well - lesson03 mailroom guide has an example of quit function that utilizes exit() function.